### ISSUE TITLE
RFC: Extend PrimitiveArray to SizedArray

### ISSUE THREAD

**Issue by TennyZhuang** [singularity-data/risingwave#2190]
Currently, we use `PrimitiveArray` to represent pg types which can also be represented by rust primitive types, e.g.: ints, floats. There are still some other fixed sized pg types, e.g. Interval, Date, Time, Timestamp, Decimal (in our implementation, it's sized now). The codes are very similar, so I prefer to extend `PrimitiveArray` to `SizedArray`, and try to reuse codes when possible.

**Comment by TennyZhuang** [comment]
Better naming is welcome!

**Comment by skyzh** [comment]
maybe FixedSizeArray?

**Comment by skyzh** [comment]
PrimitiveArray also looks good to me. Chrono date types are indeed primitives.

**Comment by TennyZhuang** [comment]
> maybe FixedSizeArray?

looks better

**Comment by fuyufjh** [comment]
"FixedSizeArray" sounds like the array is fixed-sized 🤣

**Comment by TennyZhuang** [comment]
> "FixedSizeArray" sounds like the array is fixed-sized 🤣

FixedItemArray?

**Comment by fuyufjh** [comment]
Hmmm, still sounds like the number of items is fixed...

-----

> PrimitiveArray also looks good to me. Chrono date types are indeed primitives.

+1 for this.

**Comment by fuyufjh** [comment]
After thinking for 3 mins, I have no preference between "SizedArray" and "PrimitiveArray" 😅

**Comment by TennyZhuang** [comment]
Primitive looks like the items are primitive rust types.

**Comment by xxchan** [comment]
> Rather than letting the other side of protobuf to decode `rust_decimal` format, we would like to use a neutral format - variable-length string (#557). Performance is not a concern at this stage. cc @liurenjie1024 @BowenXiao1999 

_Originally posted by @xiangjinwu in https://github.com/singularity-data/risingwave-legacy/pull/550#discussion_r694460801_

Do we make `DecimalArray` a `PrimitiveArray` now, which will change its encoding?

**Comment by TennyZhuang** [comment]
> > Rather than letting the other side of protobuf to decode `rust_decimal` format, we would like to use a neutral format - variable-length string (#557). Performance is not a concern at this stage. cc @liurenjie1024 @BowenXiao1999
> 
> _Originally posted by @xiangjinwu in [singularity-data/risingwave-legacy#550 (comment)](https://github.com/singularity-data/risingwave-legacy/pull/550#discussion_r694460801)_
> 
> Do we make `DecimalArray` a `PrimitiveArray` now, which will change its encoding?

We should do it after some necessary discussions.

**Comment by xiangjinwu** [comment]
To clarify, `DecimalArray` is also an array with elements of fixed size inside rust. It is only encoded as variable length string when communicating via protobuf.

**Comment by xxchan** [comment]
> To clarify, `DecimalArray` is also an array with elements of fixed size inside rust. It is only encoded as variable length string when communicating via protobuf.

Yes, but `PrimitiveArray` will take charge of `to_protobuf`. So if we still want such protobuf encoding for `DecimalArray`, I prefer no to change it. Otherwise we can do the change and reduce some duplicate code.

**Comment by xxchan** [comment]
I found Apache Arrow also call it PrimitiveArray 🤔 And its boolean array is also PrimitiveArray.
